### DISCUSSION TITLE
Custom ORDER BY clause

### DISCUSSION THREAD

**Discussion by Darkside73** [noam-honig/radweb#438] [General]
How to add custom ORDER BY for ordering by complex expression rather than by one field? For example sort results by full text rank in PostgreSQL:
```sql
select title from posts
order by ts_rank(textsearchable_index_col, websearch_to_tsquery('search query')) desc;
```
Or at least is there a way to extract this into entity class similar to `Filter.createCustom`?

**Comment by noam-honig** [comment]
Cool request.

You can achieve it today with an sqlExpression field (although it's not optimal) can you try that and let me know?

**Reply by Darkside73** [reply]
I've ended up with following (not optimal as well) solution:
```ts
class Post {
...
static fulltextSort = Filter.createCustom<Post, { query: string }>(
    ({ query }) =>
      SqlDatabase.rawFilter(
        ({ param }) => {
          return `ts_tank(textsearchable_index_col, websearch_to_tsquery(${param(query)}))`
        }
      )
  )
}

// use it in query builder  
// @ts-expect-error
options.orderBy![Post.fulltextSort({ query: query.q })] = 'desc'
await repo.find({ ..options })
 ```

Not sure how to use `sqlExpression` exactly

**Reply by noam-honig** [reply]
Does that work? I don't think it should, as I do not resolve the customFilter as part of the sort.

As for the sqlExpression - you are right, I was wrong - I missed the parameter coming in.

I'll see what I can do about the custom sort.

**Reply by Darkside73** [reply]
You're right. It does not. Looking forward for another approach

**Comment by noam-honig** [comment]
Hi @Darkside73, we have an idea on how we can implement this in a future version, 
We also think that the immediate next step for a custom sort - would be custom field, I'm guessing that if you want to sort addresses by distance to a point, you also want to display that distance.

in the meantime, I have a workaround - let me know what you think - here's an example implementation of the approach:
https://stackblitz.com/edit/github-r1uhb4?file=src%2Fshared%2Fmodel.ts

this shows a list of tasks, and when you type a value in the top input, it sorts the rows based on the number of occurrences of that string in each task

1. It defined a dummy field `queryArgs`, that'll be used to hold the query
2. The' backendPreprocessFilter' takes that value if received in the filter and stores it in the `remult.context` object
3. In the `sqlExpression` of the `queryCountInTitle` field it uses that value from `remult.context` to create the sql expression.
4. In the frontend that field can be used as sort, and as a field
```ts
    taskRepo
      .find({
        limit: 20,
        orderBy: { queryCountInTitle: 'desc' },
        where: {
          queryArgs: {
            query,
          },
        },
      })
```

![image](https://github.com/remult/remult/assets/16635859/61f70fad-a341-4452-9cdc-e60988fa9097)

Let me know if you can use this as in interim solution, until we'll implement this as a feature

```ts
import { Entity, Fields, dbNamesOf, remult } from 'remult'

@Entity<Task>('tasks', {
  allowApiCrud: true,
  backendPreprocessFilter: (filter) => {
    for (let and of filter?.$and || []) {
      const args = and.queryArgs as QueryArgument
      if (args?.query !== undefined) {
        // remult.context is separate for each request
        remult.context.queryArgument = args
        delete and.queryArgs
      }
    }
    return filter
  },
})
export class Task {
  @Fields.cuid()
  id = ''

  @Fields.string()
  title = ''

  @Fields.boolean()
  completed = false

  @Fields.createdAt()
  createdAt?: Date

  @Fields.integer<Task>({
    sqlExpression: async () => {
      const t = await dbNamesOf(Task)
      if (!remult.context.queryArgument?.query) return `(select 0) `
      const escapedSubstring = escapeSqlString(
        remult.context.queryArgument.query
      )
      return `(LENGTH(${t.title}) - LENGTH(REPLACE(${t.title}, '${escapedSubstring}', ''))) / LENGTH('${escapedSubstring}')  
      `
    },
  })
  queryCountInTitle = 0

  @Fields.json({
    serverExpression: () => ({}),
    valueConverter: {
      toJson: (x) => JSON.stringify(x),
      fromJson: (y) => JSON.parse(y),
    },
  })
  queryArgs?: QueryArgument
}

type QueryArgument = {
  query: string
}

//  https://remult.dev/docs/custom-options#extending-remult-s-context-property-for-request-specific-information
declare module 'remult' {
  export interface RemultContext {
    queryArgument?: QueryArgument
  }
}

function escapeSqlString(str: string) {
  if (typeof str !== 'string') {
    return str
  }

  return str
    .replace(/\\/g, '\\\\') // Escape backslashes
    .replace(/'/g, "''") // Escape single quotes
}

```

**Reply by Darkside73** [reply]
Sorry for delay. There are cases when custom ORDER BY clause is used without displaying actual values to user. Rank values from full text search is the one of those cases. I think it is better for framework's agility to simply allow the developer to use SQL like for WHERE part of the query.
I'll try you solution for my task. Thanks

**Reply by noam-honig** [reply]
I agree, we'll probably add that soon - in the meantime check out the workaround I've sent and see if it is useful as an interim solution

**Reply by Darkside73** [reply]
Tried your solution. It gives an error:
```
 error: column "queryArgs" does not exist
 ```
The model:
```ts
type QueryArgument = {
  query: string
}

function escapeSqlString (str: string) {
  if (typeof str !== 'string') {
    return str
  }

  return str
    .replace(/\\/g, '\\\\') // Escape backslashes
    .replace(/'/g, '\'\'') // Escape single quotes
}

@Entity('posts', {
  backendPreprocessFilter: (filter) => {
    for (const and of filter?.$and || []) {
      const args = and.queryArgs as QueryArgument
      if (args?.query !== undefined) {
        // remult.context is separate for each request
        remult.context.queryArgument = args
        delete and.queryArgs
      }
    }
    return filter
  }
})
export class Post {
  @Fields.cuid()
    id = ''

  // ...

  // eslint-disable-next-line no-use-before-define
  @Fields.number<Post>({
    sqlExpression: () => {
      if (!remult.context.queryArgument?.query) { return '(select 0) ' }
      const escapedSubstring = escapeSqlString(
        remult.context.queryArgument.query
      )
      return `ts_rank(textsearchable_index_col, websearch_to_tsquery('${escapedSubstring}'))`
    }
  })
    tsRank = 0

  @Fields.json({
    serverExpression: () => ({}),
    valueConverter: {
      toJson: x => JSON.stringify(x),
      fromJson: y => JSON.parse(y)
    }
  })
    queryArgs?: QueryArgument

  static fulltextSearch = Filter.createCustom<Post, { query: string }>(
    ({ query }) =>
      SqlDatabase.rawFilter(
        ({ param }) => {
          return `textsearchable_index_col @@ websearch_to_tsquery(${param(query)})`
        }
      )
  )
}
```

Usage context:
```ts
if (query.q) {
  filterOptions.where!.queryArgs = { query: query.q }
  filterOptions.where!.$and = [Post.fulltextSearch({ query: query.q })]
  filterOptions.orderBy!.tsRank = 'desc'
}
```

**Comment by noam-honig** [comment]
Can you adjust the stack blitz I sent you to see the issue?
Or if you prefer to do an online session, join our discord and I’ll be
available later this week

**Reply by Darkside73** [reply]
I've figured out. The issue appeared since I do not use filter (only custom orderBy). So I've hacked backendPreprocessFilter like that:
```ts
backendPreprocessFilter: (filter) => {
  remult.context.queryArgument = filter.queryArgs
  delete filter.queryArgs
  return filter
}
```
And now it produces desirable SQL
```sql
select
  "id", "title", ts_rank(textsearchable_index_colA, websearch_to_tsquery('query'))
from "posts"
where
  textsearchable_index_col @@ websearch_to_tsquery($1)
Order By
  ts_rank(textsearchable_index_colA, websearch_to_tsquery('query')) desc

{ arguments: { '$1': 'query' }
```

**Comment by noam-honig** [comment]
Cool - we will simplify it in the future

Thanks

**Comment by noam-honig** [comment]
Hi @Darkside73 I've started to play around with a solution to this issue - trying to keep things as tightly typed as I can - checkout this test, let me know what you think

https://github.com/remult/remult/blob/query-argumets/projects/tests/dbs/test-sql-database.spec.ts#L100-L128

**Reply by Darkside73** [reply]
As I understand it, we still need to declare the virtual field. I think it's beneficial for typing and creating a high-level model interface. However, is it possible to implement a more classic and agile ORM interface like the following?
```ts
await repo.find({
    orderBySQL: ['length(col) - length(:query)', { query: inputComesFromUser }]
})
```
It's always handy to have the ability to write raw SQL for different statement parts

**Reply by noam-honig** [reply]
Hi @Darkside73, we need a way to wrap the SQL like we do in sqlExpression or customFilter - to be able to use it also through the frontend.

So at the minimum it'll be something like:
```ts
class whatever{

static lengthSort = Sort.customSort<string>(query=>SqlDatabase.customOrderBy(()=>'length(col) - length(:query)', { query: inputComesFromUser }])
}
```

and 
```ts
await repo.find({orderBy:whatever.lengthSort(inputFromUser)})
```
Our goal is to be able to also wrap any SQL with something safe that can be called from the front end


What do you think?

**Reply by noam-honig** [reply]
To elaborate some more - to allow safe usage from the frontend, any custom SQL should be defined in the entity and only "called" when using find or similar commands - as such, customFilter is defined in the entity (and it can use `rawSql`).

I was thinking that `sqlExpression` with an argument will solve both the dynamic field and dynamic order by problems - do you think that there should be a separate solution for custom orderby, where you do not need the field or the field solution is "good enough"?

**Reply by Darkside73** [reply]
You're right. I've forgot about frontend part) Encapsulating in static property looks pretty neat. But field solution uses existing framework concepts like `sqlExpression`, so I think maybe we should stick with it
